File: lib/services/ai_service.dart
import 'dart:math';
import 'package:google_generative_ai/google_generative_ai.dart';
import 'package:flutter/foundation.dart';

class AIService {
  static Future<String> improveText(String text, {String? apiKey}) async {
    if (apiKey != null && apiKey.isNotEmpty) {
      try {
        final model = GenerativeModel(model: 'gemini-1.5-flash', apiKey: apiKey);
        final content = [Content.text('Improve the following text for a README file, making it more professional, concise, and engaging:\n\n$text')];
        final response = await model.generateContent(content);
        return response.text ?? text;
      } catch (e) {
        debugPrint('Gemini API Error: $e');
        return '$text (Error: ${e.toString().replaceAll('GenerativeAIException: ', '')})';
      }
    }

    // Mock fallback
    await Future.delayed(const Duration(milliseconds: 800));
    if (text.isEmpty) return 'Generated content based on context...';
    return '$text (AI Enhanced)';
  }

  static Future<String> generateDescription(String topic, {String? apiKey}) async {
    if (apiKey != null && apiKey.isNotEmpty) {
      try {
        final model = GenerativeModel(model: 'gemini-1.5-flash', apiKey: apiKey);
        final content = [Content.text('Generate a short, engaging project description for a project about: $topic. Include emojis where appropriate.')];
        final response = await model.generateContent(content);
        return response.text ?? 'Could not generate description.';
      } catch (e) {
        debugPrint('Gemini API Error: $e');
        return 'Error: ${e.toString().replaceAll('GenerativeAIException: ', '')}';
      }
    }

    await Future.delayed(const Duration(milliseconds: 800));
    return 'A comprehensive solution for $topic built with modern technologies.';
  }

  static Future<String> fixGrammar(String text, {String? apiKey}) async {
    if (apiKey != null && apiKey.isNotEmpty) {
      try {
        final model = GenerativeModel(model: 'gemini-1.5-flash', apiKey: apiKey);
        final content = [Content.text('Fix grammar and spelling in the following text. Preserve markdown formatting:\n\n$text')];
        final response = await model.generateContent(content);
        return response.text ?? text;
      } catch (e) {
        debugPrint('Gemini API Error: $e');
        return text;
      }
    }

    await Future.delayed(const Duration(milliseconds: 500));
    String fixed = text.trim();
    if (fixed.isNotEmpty) {
      fixed = fixed[0].toUpperCase() + fixed.substring(1);
      if (!fixed.endsWith('.') && !fixed.endsWith('!') && !fixed.endsWith('?')) fixed += '.';
    }
    return fixed;
  }

  static Future<String> generateReadmeFromStructure(String structure, {String? apiKey}) async {
    if (apiKey != null && apiKey.isNotEmpty) {
      try {
        final model = GenerativeModel(model: 'gemini-1.5-flash', apiKey: apiKey);
        final prompt = '''
You are an expert developer tool designed to generate comprehensive README.md files.
Analyze the following codebase structure and generate a professional, detailed README.md file in Markdown format.

Include:
1. Project Title and Description.
2. Key Features.
3. Tech Stack.
4. Quick Start / Installation.
5. Project Structure.
6. How to Contribute.

Output ONLY the raw Markdown content.

Structure:
$structure
''';
        final content = [Content.text(prompt)];
        final response = await model.generateContent(content);
        return response.text ?? '# Generated README';
      } catch (e) {
        debugPrint('Gemini API Error: $e');
        return '# Error: ${e.toString()}';
      }
    }
    return '# Mock README\n\nProvide an API Key to use AI generation.';
  }
}

File: src/services/AIService.js
import { GoogleGenerativeAI } from "@google/generative-ai";

class AIService {
  /**
   * Initializes the Gemini model with the provided API key.
   */
  _getModel(apiKey) {
    if (!apiKey) throw new Error("API Key is missing");
    const genAI = new GoogleGenerativeAI(apiKey);
    return genAI.getGenerativeModel({ model: "gemini-1.5-flash" });
  }

  /**
   * Improves existing text to be more professional.
   */
  async improveText(text, apiKey) {
    try {
      const model = this._getModel(apiKey);
      const prompt = `As a technical writer, improve this README section to be professional, concise, and engaging. Return only the improved text:\n\n${text}`;
      const result = await model.generateContent(prompt);
      const response = await result.response;
      return response.text() || text;
    } catch (e) {
      console.error('AI Improve Error:', e);
      return text;
    }
  }

  /**
   * Generates a short engaging description for a topic.
   */
  async generateDescription(topic, apiKey) {
    try {
      const model = this._getModel(apiKey);
      const prompt = `Generate a short engaging description for: ${topic}`;
      const result = await model.generateContent(prompt);
      const response = await result.response;
      return response.text() || `A project about ${topic}`;
    } catch (e) {
      console.error('AI Desc Error:', e);
      return `A project about ${topic}`;
    }
  }

  /**
   * Magic Compose: Generates a full ReadmeElement list based on a prompt.
   */
  async magicCompose(prompt, apiKey) {
    const model = this._getModel(apiKey);
    const aiPrompt = `
      You are an expert GitHub README generator. Based on this description: "${prompt}",
      generate a professional README structure.
      Return the response ONLY as a JSON array of objects.
      Each object must have a "type" field and data fields.
      Supported types:
      - "heading" (text, level: 1-3)
      - "paragraph" (text)
      - "list" (items: list of strings, isOrdered: bool)
      - "codeBlock" (code, language)
      - "divider" (no extra fields)

      Example: [{"type": "heading", "text": "Project Name", "level": 1}, {"type": "divider"}]
      Return ONLY the raw JSON. No markdown backticks.
    `;

    let cleanJson = '';
    try {
      const result = await model.generateContent(aiPrompt);
      const response = await result.response;
      cleanJson = response
        .text()
        .replace(/```json/g, '')
        .replace(/```/g, '')
        .trim();
    } catch (e) {
      console.error('Magic Compose Error:', e);
      throw new Error(e?.message || 'AI request failed. Please try again.');
    }

    try {
      return JSON.parse(cleanJson || '[]');
    } catch (e) {
      console.error('Magic Compose JSON Error:', e);
      throw new Error('AI returned an invalid response. Try a shorter, more specific prompt.');
    }
  }

  /**
   * Analyzes project elements and provides a health score.
   */
  calculateHealthScore(elements) {
    if (!elements || elements.length === 0) return 0.0;
    let score = 0.0;

    const hasH1 = elements.some(e => e.type === 'heading' && e.level === 1);
    const hasImage = elements.some(e => e.type === 'image');
    const hasCode = elements.some(e => e.type === 'code');
    const hasSocials = elements.some(e => e.type === 'socials');

    if (hasH1) score += 30;
    if (hasImage) score += 20;
    if (hasCode) score += 20;
    if (hasSocials) score += 10;
    if (elements.length > 5) score += 20;

    return Math.min(Math.max(score, 0), 100);
  }
}

export default new AIService();
